File: src/main/kotlin/org/kravbank/java/api/PublicationResource.java
package org.kravbank.java.api;



import org.kravbank.java.model.Publication;

import javax.transaction.Transactional;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.net.URI;
import java.util.List;

@Path("/publications")
//@Produces(APPLICATION_JSON)
//@Consumes(APPLICATION_JSON)

public class PublicationResource {

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    public Response getAll() {
        List<Publication> publications = Publication.listAll();
        return Response.ok(publications).build();
    }

    @GET
    @Path("{id}")
    @Produces(MediaType.APPLICATION_JSON)
    public Response getOne(@PathParam("id") Long id) {
        return Publication.findByIdOptional(id)
                .map(publication -> Response.ok(publication).build())
                .orElse(Response.status(Response.Status.NOT_FOUND).build());
    }

    @POST
    @Transactional
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    public Response create(Publication publication) {
        Publication.persist((publication));
        if (publication.isPersistent()) {
            return Response.created(URI.create("/publications" + publication.id)).build();
        }
        return Response.status(Response.Status.BAD_REQUEST).build();
    }

    @GET
    @Path ("publication/{date}")
    @Produces(MediaType.APPLICATION_JSON)
    public Response getByDate(@PathParam("date") String date) {
        List <Publication> projects = Publication.list("SELECT p from Publication p WHERE p.date = ?1 ORDER BY id DESC", date);
        return Response.ok(projects).build();
    }

    @GET
    @Path ("bankid/{bankid}")
    @Produces(MediaType.APPLICATION_JSON)
    public Response getByBankId(@PathParam("bankid") String bankid) {
        return Publication.find("bankId", bankid)
                .singleResultOptional()
                .map(publication -> Response.ok(publication).build())
                .orElse(Response.status(Response.Status.NOT_FOUND).build());
    }

    @DELETE
    @Path("{id}")
    @Produces(MediaType.APPLICATION_JSON)
    @Transactional
    public Response deleteById(@PathParam("id") Long id){
        boolean deleted = Publication.deleteById(id);
        if (deleted){
            return Response.noContent().build();
        }
        return Response.status(Response.Status.BAD_REQUEST).build();
    }
}

File: src/main/kotlin/org/kravbank/api/PublicationResource.kt
package org.kravbank.api;

import org.kravbank.utils.form.publication.PublicationForm
import org.kravbank.utils.form.publication.PublicationFormUpdate
import org.kravbank.service.PublicationService
import org.kravbank.utils.mapper.publication.PublicationMapper
import org.kravbank.utils.mapper.publication.PublicationUpdateMapper
import java.net.URI
import java.util.ArrayList
import javax.enterprise.context.RequestScoped
import javax.transaction.Transactional
import javax.ws.rs.*
import javax.ws.rs.core.MediaType.APPLICATION_JSON
import javax.ws.rs.core.Response

@Path("/api/v1/projects/{projectref}/publications")
@Produces(APPLICATION_JSON)
@Consumes(APPLICATION_JSON)
@RequestScoped
class PublicationResource(val publicationService: PublicationService) {
    //GET PUBLICATION
    @GET
    @Path("/{publicationref}")
    fun getPublication(
        @PathParam("projectref") projectRef: String, @PathParam("publicationref") publicationRef: String
    ): Response {
        val publication = publicationService.get(projectRef, publicationRef)
        //mapper fra entity
        val publicationDTO = PublicationMapper().fromEntity(publication)
        return Response.ok(publicationDTO).build()
    }

    //LIST PUBLICATIONS
    @GET
    fun listPublications(@PathParam("projectref") projectRef: String): Response {
        val publications = publicationService.list(projectRef)
        val publicationsDTO = ArrayList<PublicationForm>()
        //mapper fra entity
        for (n in publications) publicationsDTO.add(PublicationMapper().fromEntity(n))
        return Response.ok(publicationsDTO).build()
    }
    //CREATE PUBLICATION
    @Transactional
    @POST
    fun createPublication(@PathParam("projectref") projectRef: String, newPublication: PublicationForm): Response {
        val publication = publicationService.create(projectRef, newPublication)
        //sender ny publication ref i respons header
        return Response.created(URI.create("/api/v1/projects/$projectRef/publications/" + publication.ref)).build()
    }

    //DELETE PUBLICATION
    @DELETE
    @Path("/{publicationref}")
    @Transactional
    fun deletePublication(
        @PathParam("projectref") projectRef: String, @PathParam("publicationref") publicationRef: String
    ): Response {
        val publication = publicationService.delete(projectRef, publicationRef)
        val publicationDTO = PublicationMapper().fromEntity(publication)
        // sender slettet publication ref i body
        return Response.ok(publicationDTO.ref).build()
    }

    //UPDATE PUBLICATION
    @PUT
    @Path("{publicationref}")
    @Transactional
    fun updatePublication(
        @PathParam("projectref") projectRef: String,
        @PathParam("publicationref") publicationRef: String,
        updatedPublication: PublicationFormUpdate
    ): Response {
        val publication = publicationService.update(projectRef, publicationRef, updatedPublication)
        // mapper fra entity
        val publicationUpdateDTO = PublicationUpdateMapper().fromEntity(publication)
        return Response.ok(publicationUpdateDTO).build()
    }
}